### PR TITLE
Collision events

### PR DESCRIPTION
**Pull request by grozen**
This pull request exposes an alternative method for handling collision by binding to events that fire as collisions take place. The events in question are "HitOn" for when collisions occur and "HitOff" for when they cease.

This pull request was born as an effort to solve issue #147, which it does (as you can see in the following fiddle: http://jsfiddle.net/nm2nL/5/).

I've added some tests as well, and will more than welcome your scrutiny.

### CODE REVIEW

**Comment by starwed** [comment]
I've only glanced at it so far, but I think you're checking for hits with "EnterFrame"?

There's no guarantee that these events happen in a particular order, so it might be that an object moves _after_ the collision check occurs.  So if you try to use this to prevent overlaps, for instance, it might fail.  It would only lag by a single frame at worst, and so not be very noticeable to the eye -- but if you want to treat "no overlaps" as an invariant it might cause issues.

One solution would be to bind to "Move" and "Rotation" events instead of "EnterFrame".  But these could conceivably trigger multiple times in a physics frame, and thus be more expensive to calculate.   (I've noticed that in practice, the collision detection code is the most CPU intensive thing that happens outside of rendering.)  And they wouldn't work when another entity move on top of _you_.

I'm not really sure what the best approach is.  Maybe the best use of `onHit` isn't to prevent collisions, but just to detect when you pass over powerups, switches, and the like, in which case it doesn't really matter if it might occasionally lag by a single frame.

**Comment by starwed** [comment]
It's possible that a long term solution is to have code in the hashmap implementation that deals with this when an entities position updates.  It already does half the work in any case; it wouldn't be too crazy for it to maintain the collision status of every "Collision" element.

**Comment by grozen** [comment]
I am indeed checking for hits on _EnterFrame_, @starwed, but this is no different than what "onHit" currently does.
While it does not make your point less true, the real question is whether or not it is actually a problem. If you check out the fiddle I made then everything seems in order, as in, you'd never even suspect that the problem you describe exists. Then again, it might be pure luck.

Also, is your second comment more of a note to yourself? I am afraid I did not follow it, exactly.

**Comment by starwed** [comment]
> I am indeed checking for hits on EnterFrame, @starwed, but this is no different than what "onHit" currently does.

That's true.  As long as it's documented in some way, sounds good to me!

> Also, is your second comment more of a note to yourself? I am afraid I did not follow it, exactly. 

Sorry, just thinking out loud.  As an aid for collision detection and rendering, Crafty keeps track of the location of 2D entities in course-grained spatial index.  Every time an entity changes positions, we update it's position in that index; if we also wanted to update the collision status of entities every time they moved, that would be the obvious place to put such code.

**Comment by grozen** [comment]
Thanks for the clarification. I've gone ahead and enhanced the documentation.

File: src/collision.js [hunk 1]
```diff
@@ -313,6 +336,196 @@ Crafty.c("Collision", {
...
     },
 
+
+    /**@
+     * #._createCollisionHandler
+     * @comp Collision
+     * @sign public void .checkHits(String component, Object collisionData)
+     * @param component - The name of the component for which this handler
+     * checks for collisions.
+     * @param component# - Collision data object used to track collisions with
```
Comment (starwed) [inline src/collision.js]: Should be collisionData the second time?

**Comment by starwed** [comment]
Ok, I've gone through and the API seems sensible to me.  I haven't tested it in action yet, though!

Two things:
- Could you squash down all the commits at some point before we land it?  There are a _lot_.  :)
- The existing implementation of `hit` is inefficient if you care about multiple components.  It fetches a list of all potential collisions, then discards any potential matches that don't have the right component.  That means if you call it multiple times per frame with the same component, it'll repeat that first stage multiple times.  That said, it won't waste time on the more time consuming sort of collision check, so maybe that doesn't really matter.

Anyway, once the commits are squashed down, this can go ahead and land.  It'll probably get more of the sort of testing it needs once it's  actually in the development branch.  :)

**Comment by grozen** [comment]
Neat. Keep in mind that like my previous change the test page contains a playground that demonstrates the feature. However, it displays some very odd behavior as all the draggable component refuse to get dragged until the window is resized... (I am at a loss as to why this happens, at the moment. The problem is even more severe on chrome).

I will be sure to squash my commits before long.

Regarding the inefficiency of 'hit', I suspect this is outside the scope of this change, not that I would mind tackling this independently. Specifically, I suspect the way to go here is to have hit receive multiple components so that it can find all the interesting collisions in one pass.

**Comment by grozen** [comment]
Ok, I **think** I've performed a proper squash. I consider rebasing to be 50% dark magic.

**Comment by starwed** [comment]
Oh hmm, wasn't thinking about the tests.    I think sometime after you opened this PR, we switched to calling them as .js files for everything except the animation tests.  For the automated tests, does this need to be a standalone html file, or can it just be called from `index.html` like the others are now?

I do like the idea of the playground, though -- they're a nice addition to the automated tests.

**Comment by grozen** [comment]
All the tests need to pass is a Crafty stage, so they aren't really tied to the playground in any way. Still, I consider the playgrounds a complementary part to the tests I produce, as they allow you to get a more tangible sense of the feature being tested.

Not to mention they serve as an example of just how the feature is used outside of a test.

Seeing how we can list individual test files to be run in the `Gruntfile` instead of listing them all under index.html, is this such a bad thing?

I'd hate to be a troublemaker, obviously, and if need be I'll move the tests.

File: src/collision.js [hunk 2]
```diff
@@ -206,32 +221,40 @@ Crafty.c("Collision", {
...
      * }]
      * ~~~
-     * `MBR` is your standard axis aligned rectangle intersection (`.intersect` in the 2D component).
-     * `SAT` is collision between any convex polygon.
+     * **obj:** The entity with which the collision occured.
+     * **type:** Collision detection method used. One of:
+     * + *MBR:* Standard axis aligned rectangle intersection (`.intersect` in the 2D component).
+     * + *SAT:* Collision between any two convex polygons.
+     * **overlap:** If SAT collision was used, this will signify the overlap percentage between the colliding entities.
      *
```
Comment (mucaho) [inline src/collision.js:256]: Could you add a line here saying: "`SAT` collision can only occur if both colliders have the `Collision` component."?

**Comment by grozen** [comment]
@mucaho That's a very good point. Added.

**Comment by starwed** [comment]
@grozen the only issue with having multiples files is that it takes longer to run them, since apparently a big chunk of the time is spent setting up the framework.  (On my machine each additional test html file adds ~5s of time to `grunt check`, which obviously adds up with a lot of tests.)

Since everything else is good I'll just go ahead and merge this, and deal with the tests in a followup PR.  :)

**Comment by grozen** [comment]
@starwed How very surprising. I would not expect such a performance penalty for merely having additional test files.

I'll do my best to detach my tests from their respective playgrounds without sacrificing the playgrounds in the process.

**Comment by starwed** [comment]
@grozen  Apparently running the tests happens pretty quickly, and most of the time is spent by grunt setting up the testing framework separately for each file?

Having a separate directory for playgrounds and demo-like files would be useful.  In general, some behavior is easier to check visually than through unit testing, so it would be nice to do so before official releases.  :)

**Pull request merged**